File: Hieu/triangle.py
def tknp1(l,r,c,k):
    res = -1
    while l <= r:
        g = (l+r)//2
        if c[g] < k:
            res = g
            l = g + 1
        else:
            r = g - 1
    return res
def tknp2(l,r,c,k):
    res = -1
    while l <= r:
        g = (l+r)//2
        if c[g] > k:
            res = g
            r = g - 1
        else:
            l = g + 1
    return res
fi = open('triangle.inp','r'); fo = open('triangle.out','w')
n = int(fi.readline())
a = [0] + list(map(int,fi.readline().split()))
b = [0] + list(map(int,fi.readline().split()))
c = [0] + list(map(int,fi.readline().split()))
c.sort()
res = 0
for i in range(1,n+1):
    for j in range(1,n+1):
        vtc = tknp1(1,n,c,a[i]+b[j])
        vtd = tknp2(1,n,c,abs(a[i]-b[j]))
        if vtd <= vtc and vtd != -1 and vtc != -1:
            res += vtc - vtd + 1
fo.write(str(res))
fi.close();fo.close()

File: TRIANGLE.cpp
#include <bits/stdc++.h>
#define ll long long
#define name "TRIANGLE"
using namespace std;
int n,a[1005],b[1005],C[1005],ans,T,P;
int tknp1(ll x,ll d,ll c)
{
    ll g,kq=0;
    while(d<=c)
    {
        g=(d+c)/2;
        if (C[g]>x)
        {
            kq=g;
            c=g-1;
        }
        else
            d=g+1;
    }
    return kq;
}
int tknp2(ll x,ll d,ll c)
{
    ll g,kq=0;
    while(d<=c)
    {
        g=(d+c)/2;
        if (C[g]<x)
        {
            kq=g;
            d=g+1;
        }
        else
            c=g-1;
    }
    return kq;
}
int main()
{
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);
    //freopen(name".INP","r",stdin);
    //freopen(name".OUT","w",stdout);
    cin>>n;
    for (int i=1;i<=n;i++) cin>>a[i];
    for (int i=1;i<=n;i++) cin>>b[i];
    for (int i=1;i<=n;i++) cin>>C[i];
    sort(C+1,C+1+n);
    for (int i=1;i<=n;i++)
    {
        for (int j=1;j<=n;j++)
        {
            T=tknp1(abs(a[i]-b[j]),1,n);
            P=tknp2(a[i]+b[j],1,n);
            ans+=P-T+1;
        }
    }
    cout<<ans;
}
